sdk/attestation/Azure.Security.Attestation/src/AttestationToken.cs

```diff
@@ -143,43 +202,390 @@ protected AttestationToken()
+                {
+                    return DateTimeOffset.FromUnixTimeSeconds((long)Payload.IssuedAtTime.Value);
+                }
+                return null;
+            }
+        }
 
         /// <summary>
         /// Represents the body of the token encoded as a string.
         /// </summary>
-        public string TokenBody  { get => Encoding.UTF8.GetString(TokenBodyBytes.ToArray()); }
+        public virtual string TokenBody  { get => Encoding.UTF8.GetString(TokenBodyBytes.ToArray()); }
 
         /// <summary>
         /// Represents the body of the token encoded as a string.
         /// </summary>
-        public string TokenHeader { get => Encoding.UTF8.GetString(TokenHeaderBytes.ToArray()); }
+        public virtual string TokenHeader { get => Encoding.UTF8.GetString(TokenHeaderBytes.ToArray()); }
 
         /// <summary>
         /// Validate a JSON Web Token returned by the MAA.
+        /// <para/>
+        /// If the caller provides a set of signers, than that set of signers will be used as the complete set of candidates for signing.
+        /// If the caller does not provide a set of signers, then the <see cref="ValidateTokenAsync(TokenValidationOptions, IReadOnlyList{AttestationSigner}, CancellationToken)"/>
+        /// API will a set of callers derived from the contents of the attestation token.
         /// </summary>
+        /// <param name="options">Options used while validating the attestation token.</param>
         /// <param name="attestationSigningCertificates">Signing Certificates used to validate the token.</param>
-        /// <param name="validationCallback">User provided callback which allows the customer to validate the token.</param>
-        /// <returns></returns>
-        public virtual bool ValidateToken(IReadOnlyList<AttestationSigner> attestationSigningCertificates, Func<AttestationToken, AttestationSigner, bool> validationCallback = default)
+        /// <param name="cancellationToken">Token used to cancel this operation if necessary.</param>
+        /// <returns>true if the token was valid, false otherwise.</returns>
+        /// <exception cref="ArgumentNullException">Thrown if the signing certificates provided are invalid.</exception>
+        /// <exception cref="Exception">Thrown if validation fails.</exception>
+        public virtual async Task<bool> ValidateTokenAsync(TokenValidationOptions options, IReadOnlyList<AttestationSigner> attestationSigningCertificates, CancellationToken cancellationToken = default)
+        {
+            // Early out if the caller doesn't want us to validate the token.
+            if (!options.ValidateToken)
+            {
+                return true;
+            }
+
+            // Before we waste a lot of time, see if the token is unsecured. If it is, then validation is simple.
+            if (Header.Algorithm.Equals("none", StringComparison.OrdinalIgnoreCase))
+            {
+                if (!ValidateCommonProperties(options))
```

Comment:
If the caller has passed in `attestationSigningCertificates`, then they've indicated that they believe the JWT should be signed.  Shouldn't the logic here return `false` if the caller passed in a non-null value for `attestationSigningCertificates`?

Reply:
The customer has no way of knowing if the token is secured or not. The expectation is that the attestationSigningCertificates passed is going to come from the GetSigningCertificates API, which will always return a set of certificates.